core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -76,8 +79,12 @@ public String toString() {
     }
   }
 
-  // TODO(binfan): Add refcount to the UFS instance. Once the refcount goes to zero,
-  // we could close this UFS instance.
+  /**
+   * This map keeps track of the reference counter for an {@link UnderFileSystem}.
+   */
+  private final ConcurrentHashMap<Class<? extends UnderFileSystem>, AtomicLong> mReferenceCounter =
```

Comment:
the map should be keyed by `Key`, similar to `mUnderFileSystemMap` rather than `Class<? extends UnderFileSystem>`. The reason is, each `UnderFileSystem` is initialized with possibly different configuration after "ALLUXIO-2743". So even two under file systems share the same class, e.g., both are S3UnderFileSystem, they could be totally different as they use different credentials to access S3 buckets. Thus we only want to reuse the UFS when they are exactly the same scheme/authority and also using the same configuration to create, which is represented by class `Key`